lib/bridge/IrcBridge.js

```diff
@@ -754,6 +769,128 @@ IrcBridge.prototype.getBotClient = function(server) {
+}
+
+IrcBridge.prototype._solveVisibility = Promise.coroutine(function*() {
+    // For each room, do a big OR on all of the channels that are linked in any way
+
+    this._visibilityMap.mappings = yield this.getStore().getAllMappings();
+
+    let roomIds = Object.keys(this._visibilityMap.roomVisibilities);
+
+    // level is the current depth of recursion
+    let shouldBePrivate = (roomId, checkedChannels) => {
+        // If any channel connected to this room is +s, stop early and call it private
+
+        // List first connected
+        let channels = this._visibilityMap.mappings[roomId];
+        //      = ['localhost #channel1', 'localhost #channel2', ... ]
+
+        // Filter out already checked channels
+        channels = channels.filter((c) => checkedChannels.indexOf(c) === -1);
+
+        // No channels mapped to this roomId
+        if (!channels) {
+            return false;
+        }
+
+        let anyAreSecret = channels.any((channel) => {
+            let channelIsSecret = this._visibilityMap.channelModes[channel];
+
+            // If a channel mode is unknown, assume it is secret
+            if (typeof channelIsSecret === 'undefined') {
+                channelIsSecret = true;
+            }
+
+            return channelIsSecret;
+        });
+        if (anyAreSecret) {
+            return true;
+        }
+
+        // Otherwise, recurse with the rooms connected to each channel
+
+        // So get all the roomIds that this channel is mapped to and return whether any
+        //  are mapped to channels that are secret
+        return channels.map((channel) => {
+            return Object.keys(this._visibilityMap.mappings).filter((roomId2) => {
+                return this._visibilityMap.mappings[roomId2].indexOf(channel) !== -1;
+            });
+        }).any((roomId2) => {
+            return shouldBePrivate(roomId2, checkedChannels.concat(channels));
+        });
+    }
+
+    let roomCorrectVisibilities = roomIds.map(
+        (roomId) => {
+            return shouldBePrivate(roomId, []);
+        }
+    ).map((b) => b ? 'private' : 'public');
+
+    let roomsShouldBePrivate = roomIds.filter((roomId, index) => {
+        let isCurrentlyPrivate = this._visibilityMap.roomVisibilities[roomId] === 'private';
+        let roomShouldBePrivate = roomCorrectVisibilities[index] === 'private';
+
+        return !isCurrentlyPrivate && roomShouldBePrivate;
+    });
+
+    let roomsShouldBePublic = roomIds.filter((roomId, index) => {
+        let isCurrentlyPrivate = this._visibilityMap.roomVisibilities[roomId] === 'private';
+        let roomShouldBePublic = roomCorrectVisibilities[index] === 'public';
+
+        return isCurrentlyPrivate && roomShouldBePublic;
+    });
+
+    // Update rooms to correct visibilities
+    let cli = this.getAppServiceBridge().getClientFactory().getClientAs();
+
+    let promises = roomsShouldBePrivate.map((roomId) => {
+        return cli.setRoomDirectoryVisibility(roomId, "private");
```

Comment:
The fun bit is that this will race a lot because we're rescanning the entire set of mappings every single time `_solveVisibility` is called, and this is currently triggered on *any* mode change. This will result in spamming Matrix a lot, even *with* the cache (because there is a window from determining "I should change the directory visibility" to "I have received the new directory visibility and updated the cache").